allennlp/training/learning_rate_schedulers.py

```diff
@@ -118,21 +124,100 @@ def step_batch(self, epoch=None):
         for param_group, learning_rate in zip(self.optimizer.param_groups, self.get_lr()):
             param_group['lr'] = learning_rate
 
-
     def get_lr(self):
         step = max(self.last_epoch, 1)
         scale = self.factor *  (self.model_size ** (-0.5) *
                                 min(step ** (-0.5), step * self.warmup_steps ** (-1.5)))
 
         return [scale for _ in range(len(self.base_lrs))]
 
+
+class CosineWithRestarts(torch.optim.lr_scheduler._LRScheduler):  # pylint: disable=protected-access
+    """
+    Cosine annealing with restarts.
+
+    This is decribed in the paper https://arxiv.org/abs/1608.03983.
+
+    Parameters
+    ----------
+    optimizer : ``torch.optim.Optimizer``
+
+    t_max : ``int``
+        The maximum number of iterations within the first cycle.
+
+    eta_min : ``float``, optional (default=0)
+        The minimum learning rate.
+
+    last_epoch : ``int``, optional (default=-1)
+        The index of the last epoch. This is used when restarting.
+
+    factor : ``float``, optional (default=1)
+        The factor by which the cycle length (``T_max``) increases after each restart.
+
+    """
+
+    def __init__(self,
+                 optimizer: torch.optim.Optimizer,
+                 t_max: int,
+                 eta_min: float = 0.,
+                 last_epoch: int = -1,
+                 factor: float = 1.) -> None:
+        assert t_max > 0
+        assert eta_min >= 0
+        if t_max == 1 and factor == 1:
+            logger.warning("Cosine annealing scheduler will have no effect on the learning "
+                           "rate since T_max = 1 and factor = 1.")
+        self.t_max = t_max
+        self.eta_min = eta_min
+        self.factor = factor
+        self._last_restart: int = 0
```

Comment:
Sorry for interrupt, but I don't understand what is the meaning ": int"? I never saw it in python code. Could you give me some hints?

Reply:
That is a type annotation.